src/Adapters/WebForms/Controls/WebFormsLinkUtils.cs

```diff
@@ -0,0 +1,132 @@
+
+        private static IDictionary<string, ValueOrBinding<object>> BuildParameters(IDotvvmRequestContext context, RouteLinkCapability routeLinkCapability, Route webFormsRoute)
+        {
+            var parameters = webFormsRoute.Defaults?.ToDictionary(t => t.Key, t => ValueOrBinding<object>.FromBoxedValue(t.Value))
+                             ?? new Dictionary<string, ValueOrBinding<object>>();
+            foreach (var param in context.Parameters)
+            {
+                parameters[param.Key] = ValueOrBinding<object>.FromBoxedValue(param.Value);
+            }
+
+            foreach (var item in routeLinkCapability.Params)
+            {
+                parameters[item.Key] = item.Value;
+            }
+
+            return parameters;
+        }
+
+        private static string EvaluateRouteUrl(DotvvmControl container, Route webFormsRoute, IDictionary<string, ValueOrBinding<object>> parameters, RouteLinkCapability routeLinkCapability)
+        {
+            // evaluate bindings on server
+            var routeValues = new RouteValueDictionary();
+            foreach (Match param in Regex.Matches(webFormsRoute.Url, @"\{([^{}/]+)\}"))       // https://referencesource.microsoft.com/#System.Web/Routing/RouteParser.cs,48
+            {
+                var paramName = param.Groups[1].Value;
+                parameters.TryGetValue(paramName, out var value);
+                routeValues[paramName] = value.Evaluate(container) ?? "";
+            }
+
+            // generate the URL
+            return "~/"
+                   + webFormsRoute.GetVirtualPath(HttpContext.Current.Request.RequestContext, routeValues)?.VirtualPath
+                   + UrlHelper.BuildUrlSuffix(routeLinkCapability.UrlSuffix?.Evaluate(container), routeLinkCapability.QueryParameters.OrderBy(p => p.Key).Select(p => new KeyValuePair<string, object>(p.Key, p.Value.Evaluate(container))));
+        }
+
+        private static string GenerateRouteUrlExpression(DotvvmControl container, Route webFormsRoute, IDictionary<string, ValueOrBinding<object>> parameters)
+        {
+            var parametersExpression = parameters
+                .OrderBy(p => p.Key)
+                .Select(p => $"{KnockoutHelper.MakeStringLiteral(p.Key)}: {p.Value.GetJsExpression(container)}")
+                .StringJoin(",");
+            var routeUrlExpression = $"dotvvm.buildRouteUrl({KnockoutHelper.MakeStringLiteral(webFormsRoute.Url)}, {{{parametersExpression}}})";
+            return routeUrlExpression;
+        }
+
+        private static string GenerateUrlSuffixExpression(DotvvmControl container, RouteLinkCapability routeLinkCapability)
+        {
+            var urlSuffixBase = routeLinkCapability.UrlSuffix?.GetJsExpression(container) ?? "\"\"";
+            var queryParams = routeLinkCapability.QueryParameters
+                .OrderBy(p => p.Key)
+                .Select(p => $"{KnockoutHelper.MakeStringLiteral(p.Key.ToLowerInvariant())}: {p.Value.GetJsExpression(container)}")
```

Comment:
Are you sure we want to lowercase all query parameters? Those might be case sensitive.